### PR TITLE
Storage/market fix

### PR DESCRIPTION
**Pull request by Afohtim**

### CODE REVIEW

File: Assets/Scripts/UI/MarketMenu.cs [hunk 1]
```diff
@@ -96,7 +97,7 @@ public void SetMenuToClose()
 
     private void UpdateUIValues()
     {
-        float tradeRate = market.GetResourceTradeRate((GameResourceType) currentBuyItemIndex, (GameResourceType) currentSellItemIndex);
+        float tradeRate = market.GetResourceTradeRate((GameResourceType)currentSellItemIndex, (GameResourceType) currentBuyItemIndex);
```
Comment (bfilonenko) [inline Assets/Scripts/UI/MarketMenu.cs]:
```suggestion
        float tradeRate = market.GetResourceTradeRate((GameResourceType) currentSellItemIndex, (GameResourceType) currentBuyItemIndex);
```

File: Assets/Scripts/UI/MarketMenu.cs [hunk 2]
```diff
@@ -81,6 +81,7 @@ public void UpdateSellValueInputField(string text)
     public void Trade()
     {
         // TODO: call trade from market
+        market.Trade(new GameResource((GameResourceType)currentSellItemIndex, currentSellValue), new GameResource((GameResourceType)currentBuyItemIndex, currentBuyValue));
```
Comment (bfilonenko) [inline Assets/Scripts/UI/MarketMenu.cs]:
```suggestion
        market.Trade(new GameResource((GameResourceType) currentSellItemIndex, currentSellValue), new GameResource((GameResourceType) currentBuyItemIndex, currentBuyValue));
```

File: Assets/Scripts/AIForNPC/AIForNPC.cs [hunk 3]
```diff
@@ -78,6 +81,10 @@ private void Start()
         hexCellInfos = worldMap.GetHexInfoRings(transform.position, 1, colonyRaduis);
         UpdateSeed();
         Utils.Shuffle(hexCellInfos, currentSeed);
+
+
+        GameObject globalStorage = GameObject.Find("GlobalStorage");
+        id = globalStorage.GetComponent<GlobalStorage>().AddNewPlayer();
```
Comment (SymphonyOfTranquility) [inline Assets/Scripts/AIForNPC/AIForNPC.cs]:
```suggestion
        id = FindObjectOfType<GlobalStorage>().AddNewPlayer();
```

File: Assets/Scripts/Market/Market.cs [hunk 4]
```diff
@@ -7,6 +7,7 @@ public class Market : MonoBehaviour
 {
     private GameResourcesStorage initialMarketStorage;
     private GameResourcesStorage marketStorage;
+    private GameObject globalStorage;
```
Comment (bfilonenko) [inline Assets/Scripts/Market/Market.cs]:
```suggestion
    private GlobalStorage globalStorage; 
```

File: Assets/Scripts/Market/Market.cs [hunk 5]
```diff
@@ -32,6 +33,8 @@ void Start()
             prices[i] = 500;
             initialPrices[i] = 500;
         }
+
+        globalStorage = GameObject.Find("GlobalStorage");
```
Comment (bfilonenko) [inline Assets/Scripts/Market/Market.cs]:
```suggestion
        globalStorage = FindObjectOfType<GlobalStorage>();
```

File: Assets/Scripts/Market/Market.cs [hunk 6]
```diff
@@ -69,19 +73,25 @@ public float GetAvailableAmount(GameResourceType res)
...
+    public bool Trade(GameResource res1, GameResource res2, int traderId=-1)
     {
         if (needsToRefresh)
             return false;
 
         tradeMutex.WaitOne();
 
+        if (traderId == -1)
+        {
+            traderId = globalStorage.GetComponent<GlobalStorage>().CurrentMainPlayerId();
```
Comment (bfilonenko) [inline Assets/Scripts/Market/Market.cs]:
```suggestion
            traderId = globalStorage.CurrentMainPlayerId();
```

File: Assets/Scripts/Market/Market.cs [hunk 7]
```diff
@@ -69,19 +73,25 @@ public float GetAvailableAmount(GameResourceType res)
...
+        }
+
         bool isSuccessfullTrade = false;
-        if (res1.amount > marketStorage.getResourceAmount(res1.resourceType) &&
-           res2.amount > marketStorage.getResourceAmount(res2.resourceType))
+        if (res2.amount <= marketStorage.getResourceAmount(res2.resourceType))
         {
-            marketStorage.RemoveResource(res1.resourceType, res1.amount);
+            globalStorage.GetComponent<GlobalStorage>().RemoveResource(res1.resourceType, res1.amount);
```
Comment (bfilonenko) [inline Assets/Scripts/Market/Market.cs]:
```suggestion
            globalStorage.RemoveResource(res1.resourceType, res1.amount);
```

File: Assets/Scripts/Market/Market.cs [hunk 8]
```diff
@@ -69,19 +73,25 @@ public float GetAvailableAmount(GameResourceType res)
...
+            marketStorage.AddResource(res1.resourceType, res1.amount);
             marketStorage.RemoveResource(res2.resourceType, res2.amount);
+            globalStorage.GetComponent<GlobalStorage>().AddResource(res2.resourceType, res2.amount);
```
Comment (bfilonenko) [inline Assets/Scripts/Market/Market.cs]:
```suggestion
            globalStorage.AddResource(res2.resourceType, res2.amount);
```

File: Assets/Scripts/Objects/GlobalStorage.cs [hunk 9]
```diff
@@ -4,13 +4,15 @@
 
 public class GlobalStorage : MonoBehaviour
 {
-    private GameResourcesStorage storage;
+    private List<GameResourcesStorage> userStorages;
     private GameObject personalResourcesUI;
```
Comment (bfilonenko) [inline Assets/Scripts/Objects/GlobalStorage.cs]:
```suggestion
    private PersonalResourcesMenu personalResourcesUI;
```

File: Assets/Scripts/Objects/GlobalStorage.cs [hunk 10]
```diff
@@ -4,13 +4,15 @@
...
     public float refreshTime = 2;
     private float timeSinceRefresh = 0;
+    private int currentMainPlayer = 0;
     void Start()
     {
-        storage = new GameResourcesStorage();
+        userStorages = new List<GameResourcesStorage>();
+        userStorages.Add(new GameResourcesStorage(20));
         personalResourcesUI = GameObject.Find("GameMenuUI/Canvas/NoClosableMenus/PersonalResourcesMenu/Panel/PersonalResourcesMenu");
```
Comment (bfilonenko) [inline Assets/Scripts/Objects/GlobalStorage.cs]:
```suggestion
        personalResourcesUI = FindObjectOfType<PersonalResourcesMenu>();
```

File: Assets/Scripts/Objects/GlobalStorage.cs [hunk 11]
```diff
@@ -20,17 +22,67 @@ void Update()
         if (timeSinceRefresh > refreshTime)
         {
             timeSinceRefresh -= refreshTime;
-            personalResourcesUI.GetComponent<PersonalResourcesMenu>().UpdateResources(storage);
+            personalResourcesUI.GetComponent<PersonalResourcesMenu>().UpdateResources(userStorages[currentMainPlayer]);
```
Comment (bfilonenko) [inline Assets/Scripts/Objects/GlobalStorage.cs]:
```suggestion
            personalResourcesUI.UpdateResources(userStorages[currentMainPlayer]);
```

File: Assets/Scripts/AIForNPC/AIForNPC.cs [hunk 12]
```diff
@@ -19,6 +19,8 @@ public class AIForNPC : MonoBehaviour
     private float previousTime = 0f;
     private int currentSeed = 0;
 
+    public int id;
```
Comment (SymphonyOfTranquility) [inline Assets/Scripts/AIForNPC/AIForNPC.cs]: If id has to be public: add [HideInInspector] attribute and move to part with public variables.
If id has to be private: add private modifier

**Review by SymphonyOfTranquility — CHANGES_REQUESTED** [review]

**Review by bfilonenko — CHANGES_REQUESTED** [review]
Should be better to use FindObjectOfType<T> instead of GameObject.Find();
Should be better to save the link on the MonoBehaviour script than on the GameObject and after every time call GetComponent<T>()

**Review by bfilonenko — APPROVED** [review]

**Review by SymphonyOfTranquility — APPROVED** [review]

**Pull request merged**